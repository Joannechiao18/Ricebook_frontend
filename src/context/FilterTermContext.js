// context/FilterTermContext.js
import { createContext, useState } from "react";

export const FilterTermContext = createContext({
  filterTerm: "",
  setFilterTerm: () => {},
});

export const FilterTermProvider = ({ children }) => {
  const [filterTerm, setFilterTerm] = useState("");

  return (
    <FilterTermContext.Provider value={{ filterTerm, setFilterTerm }}>
      {children}
    </FilterTermContext.Provider>
  );
};
