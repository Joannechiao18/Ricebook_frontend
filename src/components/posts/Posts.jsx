import React, { useEffect, useState, useContext, useMemo, useRef } from "react";
import "./posts.scss";
import Post from "../post/Post";
import Share from "../share/Share";
import { useSelector, useDispatch } from "react-redux";
import { selectUser } from "../../reducer/authReducer";
import { setPosts } from "../../actions/postsActions";
import { selectPosts } from "../../reducer/postsReducer";
import { selectFollowedUsers } from "../../reducer/followedUsersReducer";
import { FilterTermContext } from "../../context/FilterTermContext";
import styled from "styled-components";

const BaseButton = styled.button`
  border: none;
  border-radius: 5px;
  padding: 5px 15px;
  font-size: 13px;
  cursor: pointer;
  transition: background-color 0.2s ease, color 0.2s ease;
  font-weight: 600;
  font-family: -apple-system, BlinkMacSystemFont, "Segoe UI", Roboto,
    "Helvetica Neue", Arial, "Noto Sans", sans-serif, "Apple Color Emoji",
    "Segoe UI Emoji", "Segoe UI Symbol", "Noto Color Emoji";
`;

const PageButton = styled(BaseButton)`
  color: black;
  background-color: #e3e0e0;
  margin-right: 5px;
  margin: 0 10px; // Adds 10px margin on both sides

  &:hover {
    background-color: #dedada;
  }
`;

const PaginationContainer = styled.div`
  display: flex;
  justify-content: center;
  align-items: center;
  margin-top: 20px;
`;

const PageIndicator = styled.span`
  margin: 0 10px; // Maintains the margin
  font-size: 12px; // Sets the font size
  color: #6c757d; // Bootstrap's .text-muted color, adjust as needed
`;

const Posts = () => {
  const currentUser = useSelector(selectUser);
  const posts = useSelector(selectPosts);
  const dispatch = useDispatch();
  const [currentPage, setCurrentPage] = useState(1);
  const [totalPages, setTotalPages] = useState(1); // Assuming total pages info is available from server
  const followedUsers = useSelector(selectFollowedUsers);
  const { filterTerm = "" } = useContext(FilterTermContext);
  const [fetchCount, setFetchCount] = useState(0);

  const fetchCurrentUserPosts = async () => {
    try {
      const response = await fetch(
        `https://ricebookserveryw187-8fbcb305db50.herokuapp.com/articles?username=${currentUser.username}&page=${currentPage}`,
        {
          method: "GET",
          credentials: "include",
          headers: {
            "Content-Type": "application/json",
          },
        }
      );

      const data = await response.json();

      console.log("data", data.articles.length);

      dispatch(setPosts(data.articles));
      //setTotalPages(data.articles.length); // Set the total pages (assuming the server sends this info)
      setTotalPages(data.articles.length);
    } catch (error) {
      console.error("Error fetching user posts:", error);
    }
  };

  useEffect(() => {
    fetchCurrentUserPosts();
  }, [currentPage, followedUsers]); // Dependency on currentPage

  const sortedAndFilteredPosts = useMemo(() => {
    return (Array.isArray(posts) ? posts : [])
      .filter((post) => {
        const lowercasedFilterTerm = filterTerm.toLowerCase();
        return (
          (post.author &&
            post.author.toLowerCase().includes(lowercasedFilterTerm)) ||
          (post.text && post.text.toLowerCase().includes(lowercasedFilterTerm))
          // Removed the filters for date and customId
        );
      })
      .sort((a, b) => new Date(b.date) - new Date(a.date));
  }, [posts, filterTerm]);

  return (
    console.log(totalPages),
    (
      <div className="posts container mt-5">
        <Share />
        <div className="row">
          {/* Posts mapping */}
          {sortedAndFilteredPosts.map((post) => (
            <div className="col-12 mb-5" key={post.customId}>
              <Post post={post} />
            </div>
          ))}
        </div>
        <PaginationContainer>
          <PageButton
            onClick={() => setCurrentPage((prev) => Math.max(prev - 1, 1))}
            disabled={currentPage === 1}
          >
            Previous
          </PageButton>
          <PageIndicator>
            Page {currentPage} of {totalPages}
          </PageIndicator>
          <PageButton
            onClick={() =>
              setCurrentPage((prev) => Math.min(prev + 1, totalPages))
            }
            disabled={currentPage === totalPages}
          >
            Next
          </PageButton>
        </PaginationContainer>
      </div>
    )
  );
};

export default Posts;
