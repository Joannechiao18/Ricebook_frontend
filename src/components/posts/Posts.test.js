import React from "react";
import { render, act } from "@testing-library/react";
import { Provider } from "react-redux";
import configureStore from "redux-mock-store";
import Posts from "../posts/Posts";
import fetchMock from "jest-fetch-mock";
import { FilterTermContext } from "../../context/FilterTermContext"; // Import FilterTermContext
import {
  addFollowedUser,
  followedUsersReducer,
  removeFollowedUser,
} from "../../reducer/followedUsersReducer"; // Replace with the correct path

// Mocking the Redux store
const mockStore = configureStore([]);

describe("<Posts />", () => {
  let store;

  beforeAll(() => {
    fetchMock.enableMocks();
  });

  afterAll(() => {
    fetchMock.disableMocks();
  });

  beforeEach(() => {
    store = mockStore({
      auth: {
        currentUser: {
          id: 1,
          username: "testuser",
          email: "test@example.com",
          phone: "1234567890",
          address: {
            street: "password",
            zipcode: "12345",
          },
        },
      },
      posts: [],
      followedUsers: [],
    });

    // Reset fetch mocks
    fetchMock.resetMocks();
    fetchMock.enableMocks();

    // Mock the fetch for user details and posts
    fetchMock.mockResponses(
      // Mock response for user details
      JSON.stringify({
        id: 1,
        username: "testuser",
        email: "test@example.com",
        phone: "1234567890",
        address: {
          street: "password",
          zipcode: "12345",
        },
      }),
      JSON.stringify([
        {
          id: 1,
          userId: 1,
          title: "test title",
          body: "test body",
        },
        {
          id: 2,
          userId: 1,
          title: "search title",
          body: "search body",
        },
        // ... add other mock posts as necessary
      ])
    );
  });

  it("fetches all articles for current logged in user and sets posts state", async () => {
    await act(async () => {
      render(
        <Provider store={store}>
          <Posts />
        </Provider>
      );
    });

    // Check if fetch was called
    expect(fetchMock).toHaveBeenCalledTimes(2);

    // Check the URL of the fetch call for user details
    expect(fetchMock.mock.calls[0][0]).toBe(
      "https://jsonplaceholder.typicode.com/users/1"
    );

    // Check the URL of the fetch call for posts
    expect(fetchMock.mock.calls[1][0]).toBe(
      "https://jsonplaceholder.typicode.com/posts?userId=1"
    );

    // Check if the correct action was dispatched to the store
    const actions = store.getActions();

    expect(actions[0].type).toBe("SET_POSTS");

    // Modify the test assertion to expect an array of objects with specific properties
    const expectedPayload = [
      expect.objectContaining({
        userId: 1,
        username: "testuser",
        title: "test title",
        desc: "test title",
      }),
      expect.objectContaining({
        userId: 1,
        username: "testuser",
        title: "search title",
        desc: "search title",
      }),
      // Add other expected objects as necessary
    ];

    expect(actions[0].payload).toEqual(expect.arrayContaining(expectedPayload));
  });

  it("fetches subset of articles for current logged in user given search keyword", async () => {
    await act(async () => {
      render(
        <Provider store={store}>
          <FilterTermContext.Provider
            value={{ filterTerm: "", setFilterTerm: jest.fn() }}
          >
            {" "}
            {/* Provide FilterTermContext */}
            <Posts />
          </FilterTermContext.Provider>
        </Provider>
      );
    });

    // Check if fetch was called
    expect(fetchMock).toHaveBeenCalledTimes(2);

    // Check if the correct action was dispatched to the store
    const actions = store.getActions();
    expect(actions[0].type).toBe("SET_POSTS");

    // Check if the posts are filtered based on the search keyword
    const filteredPosts = actions[0].payload.filter((post) =>
      post.body.includes("search")
    );
    expect(filteredPosts.length).toBe(1);
    expect(filteredPosts[0].body).toBe("search body");
  });

  it("should remove articles when removing a follower (posts state is smaller)", async () => {
    let actions; // Declare actions only once.

    // Initial render with a followed user
    await act(async () => {
      render(
        <Provider store={store}>
          <Posts />
        </Provider>
      );
    });

    // Check for initial fetch calls
    expect(fetchMock).toHaveBeenCalledTimes(2);

    // Mock the fetch for user details and posts for the followed user
    fetchMock.mockResponses(
      // Mock response for user details of the followed user
      JSON.stringify({
        id: 1,
        username: "testuser",
        email: "test@example.com",
        phone: "1234567890",
        address: {
          street: "password",
          zipcode: "12345",
        },
      }),
      // Mock response for posts of the followed user
      JSON.stringify([
        {
          id: 1,
          userId: 1,
          title: "current user title",
          body: "current user post body",
        },
      ]),
      JSON.stringify({
        id: 2,
        username: "follower_username",
        email: "follower_test@example.com",
        phone: "1234567890",
        address: {
          street: "password",
          zipcode: "12345",
        },
      }),
      // Mock response for posts of the followed user
      JSON.stringify([
        {
          id: 2,
          userId: 2,
          title: "follower post title",
          body: "follower post body",
        },
      ])
    );

    // Dispatch the action to add a follower
    await act(async () => {
      store.dispatch(
        removeFollowedUser({ id: 2, username: "follower_username" })
      );
    });

    // After dispatching the action to add a follower
    actions = store.getActions();

    const removeFollowerAction = actions.find(
      (action) => action.type === "REMOVE_FOLLOWED_USER"
    );

    // Check if the action to remove the follower was dispatched
    expect(removeFollowerAction).toBeDefined();

    // Remove the follower from the followedUsers state in the mock store
    const newState = followedUsersReducer(
      store.getState().followedUsers,
      removeFollowerAction
    );
    store = mockStore({ ...store.getState(), followedUsers: newState });

    expect(store.getState().followedUsers).not.toContainEqual({
      id: 2,
      username: "follower_username",
    });

    // Re-render the Posts component after removing the follower
    await act(async () => {
      render(
        <Provider store={store}>
          <Posts />
        </Provider>
      );
    });

    // Wait for promises to resolve. This ensures that all fetch calls are made.
    await Promise.resolve();

    // Fetch for the follower should have been called
    expect(fetchMock).toHaveBeenCalledTimes(4); // Two more fetches

    const userDetailsCallsForRemovedUser = fetchMock.mock.calls.filter(
      (call) => call[0] === "https://jsonplaceholder.typicode.com/users/2"
    );
    expect(userDetailsCallsForRemovedUser.length).toBe(0);

    const postCallsForRemovedUser = fetchMock.mock.calls.filter(
      (call) =>
        call[0] === "https://jsonplaceholder.typicode.com/posts?userId=2"
    );
    expect(postCallsForRemovedUser.length).toBe(0);

    // Check if the correct action was dispatched to the store
    actions = store.getActions(); // Assign new actions after dispatching

    const setPostsActions = actions.filter(
      (action) => action.type === "SET_POSTS"
    );

    const hasNewUserPosts = setPostsActions[0].payload.some(
      (post) => post.userId === 2
    ); // Assuming 2 is the userId of the new user.
    expect(hasNewUserPosts).toBe(false);
  });

  it("should add articles when adding a follower (posts state is larger)", async () => {
    let actions; // Declare actions only once.

    // Initial render with no followed users
    await act(async () => {
      render(
        <Provider store={store}>
          <Posts />
        </Provider>
      );
    });

    // Check for initial fetch calls
    expect(fetchMock).toHaveBeenCalledTimes(2);

    // Mock the fetch for user details and posts for the new follower
    fetchMock.mockResponses(
      // Mock response for user details of the new follower
      JSON.stringify({
        id: 2,
        username: "follower_username",
        email: "follower_test@example.com",
        phone: "1234567890",
        address: {
          street: "password",
          zipcode: "12345",
        },
      }),
      // Mock response for posts of the new follower
      JSON.stringify([
        {
          id: 3,
          userId: 2,
          title: "follower post title",
          body: "follower post body",
        },
      ]), // Mock response for user details of the new follower
      JSON.stringify({
        id: 1,
        username: "testuser",
        email: "test@example.com",
        phone: "1234567890",
        address: {
          street: "password",
          zipcode: "12345",
        },
      }),
      JSON.stringify([
        {
          id: 1,
          userId: 1,
          title: "test title",
          body: "test body",
        },
        {
          id: 2,
          userId: 1,
          title: "search title",
          body: "search body",
        },
        // ... add other mock posts as necessary
      ])
    );

    // Dispatch the action to add a follower
    await act(async () => {
      store.dispatch(
        addFollowedUser({
          id: 2,
          username: "follower_username",
          email: "follower_test@example.com",
          phone: "1234567890",
          address: {
            street: "password",
            zipcode: "12345",
          },
        })
      );
    });

    // After dispatching the action to add a follower
    actions = store.getActions();

    const addFollowerAction = actions.find(
      (action) => action.type === "ADD_FOLLOWED_USER"
    );

    const newState = followedUsersReducer(
      store.getState().followedUsers,
      addFollowedUser({ id: 2, username: "follower_username" })
    );
    store = mockStore({ ...store.getState(), followedUsers: newState });

    expect(actions.some((action) => action.type === "ADD_FOLLOWED_USER")).toBe(
      true
    );

    // Check the updated followedUsers state in the mock store
    expect(store.getState().followedUsers).toContainEqual({
      id: 2,
      username: "follower_username",
    });

    await act(async () => {
      render(
        <Provider store={store}>
          <Posts />
        </Provider>
      );
    });

    // Wait for promises to resolve. This ensures that all fetch calls are made.
    await Promise.resolve();

    // Fetch for the follower should have been called
    expect(fetchMock).toHaveBeenCalledTimes(6); // Four more fetches

    // Check the URL of the fetch call for user details of the new follower
    expect(fetchMock.mock.calls[2][0]).toBe(
      "https://jsonplaceholder.typicode.com/users/2"
    );

    // Check the URL of the fetch call for posts of the new follower
    expect(fetchMock.mock.calls[3][0]).toBe(
      "https://jsonplaceholder.typicode.com/posts?userId=2"
    );

    // Check if the correct action was dispatched to the store
    actions = store.getActions(); // Assign new actions after dispatching
    const setPostsActions = actions.filter(
      (action) => action.type === "SET_POSTS"
    );

    expect(setPostsActions[0].payload.length).toBe(3);

    const hasNewUserPosts = setPostsActions[0].payload.some(
      (post) => post.userId === 2
    ); // Assuming 2 is the userId of the new user.
    expect(hasNewUserPosts).toBe(true);
  });

  it("matches the snapshot", async () => {
    const { asFragment } = await act(async () => {
      return render(
        <Provider store={store}>
          <Posts />
        </Provider>
      );
    });

    expect(asFragment()).toMatchSnapshot();
  });
});
