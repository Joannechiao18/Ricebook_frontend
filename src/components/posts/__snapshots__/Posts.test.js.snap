// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Posts /> matches the snapshot 1`] = `
<DocumentFragment>
  <div
    class="posts container mt-5"
  >
    <div
      class="share"
    >
      <div
        class="container"
      >
        <div
          class="top"
        >
          <img
            alt=""
          />
          <input
            placeholder="What's on your mind, testuser?"
            type="text"
            value=""
          />
        </div>
        <hr />
        <div
          class="bottom"
        >
          <div
            class="left"
          >
            <input
              id="file"
              style="display: none;"
              type="file"
            />
            <label
              for="file"
            >
              <div
                class="item"
              >
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                  data-testid="PhotoCameraIcon"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <circle
                    cx="12"
                    cy="12"
                    r="3.2"
                  />
                  <path
                    d="M9 2 7.17 4H4c-1.1 0-2 .9-2 2v12c0 1.1.9 2 2 2h16c1.1 0 2-.9 2-2V6c0-1.1-.9-2-2-2h-3.17L15 2H9zm3 15c-2.76 0-5-2.24-5-5s2.24-5 5-5 5 2.24 5 5-2.24 5-5 5z"
                  />
                </svg>
                <span>
                  Add Image
                </span>
                 
              </div>
            </label>
          </div>
          <div
            class="right"
          >
            <button
              class="sc-dmyDmy sc-eDLKEg fLZzbZ cCJcPf"
            >
              Post
            </button>
            <button
              class="sc-dmyDmy sc-hLQTFJ fLZzbZ gWzgGU"
            >
              Cancel
            </button>
          </div>
        </div>
      </div>
    </div>
    <div
      class="row"
    />
  </div>
</DocumentFragment>
`;
